### ISSUE TITLE
Search (and replace) does not accept national characters

### ISSUE THREAD

**Issue by dpeterc** [unixwork/xnedit#25]
Main text window accepts national characters normally and displays them in XFT font using UTF8 encoding.
But If I try to enter national characters in search field using (Ctrl+F or Ctrl+R), I can not enter them.
So search and replace is significantly crippled.

**Comment by unixwork** [comment]
Yes that is a big issue. However if you are using an utf8 locale, you should be able to type characters, if they are available in the widgets font (Sans). If you want to insert a broader range of characters (like Chinese) there is a workaround to not use XFT text rendering for UI elements (the main text widget will still use XFT). 

The only good solution is to replace all Motif text input widgets with a newly written widget. This will be a lot of work, but I'm definitely planing to do that. Hopefully I have some more time to improve XNEdit in that direction in the near future.

**Comment by dpeterc** [comment]
In my own software, I use standard Motif XmTextField and XmText with XFT/UTF8, and I can both write or cut&paste national characters. All characters which are available in chosen font are displayed normally, while others are displayed as rectangle, but the contents is intact, and functionality is not affected.
If I write or paste text in national characters in main xnedit window, and then use cut and paste (Ctrl+x, ctrl+v) into search window, the national characters are filtered out in the paste. But if i do cut and paste with the secondary selection (using left mouse button to select, middle mouse button to copy), then text is pasted correctly put to (including) first national character. If I then paste same contents again with middle mouse button, then the rest of the selection is also pasted correctly. So the search text widget is actually able to display the national characters, there must be some problem in handling text selections during cut and paste.

**Comment by unixwork** [comment]
Interesting. I tried to reproduce it, however it never works like you described it.

With LANG=de_DE.UTF-8 I can write nation characters in the search window normally, also copy&paste and the primary selection works.

With LANG=de_DE I can't input national characters manually in the search window, however with copy&paste or primary selection, they are inserted invisible. Repeatedly insert with middle mouse button doesn't change anything.

Maybe I can try it with a different setup. Which OS and motif version are you using?

**Comment by dpeterc** [comment]
I am using SuSE Linux 15.0 with latest Motif snapshot from here
https://sourceforge.net/projects/motif/
+ the patch for faster XFT rendering of labels in Motif
http://bugs.motifzone.com/show_bug.cgi?id=1715

I also use MacOS Catalina + latest XQuartz (2.7.11) with Homebrew.
Same version of Motif as on Linux.
Behavior is the same.

**Comment by unixwork** [comment]
I was not really able to reproduce the problem. It highly depends on the environment. I think with an UTF-8 locale, it always works.

However, it is bad to depend on specific environment settings and the Motif textfield has other problems, therefore I started to write my own textfield widget. A first semi functional version is now available. 

To use the new textfield, add -DXNE_TEXTFIELD to your CFLAGS. This will replace the motif text widget in the find/replace dialog and in the Incremental Search Line. With that, input of all kind of unicode characters should work, regardless of locale settings. It also supports fallback fonts, so all characters should be displayed.

What's missing yet is support for primary selection and various keyboard shortcuts. Also some styling settings are not configurable yet.

Feel free to tell me which essential features are missing, then I can prioritize that.

**Comment by dpeterc** [comment]
Thanks for implementing this feature.
At a first glance, it works and I can type national characters, use Cut & Paste of such strings into search field.
Search itself is also working :-)

Things that do not work for my typical usage:
Middle mouse button (secondary selection) to paste text into search field. That is very important, since it allows you to cut and paste just with mouse. Pasting selected text out of text field with middle mouse button also dos not work, although this is less important for search text field.
Selecting parts of text by double click (whole word) and triple click (whole text) does not work.
Drag and drop with middle mouse button also does not work. (not so important to me)

Normal text edit shortcut keys to position the cursor do not work:
[Home]
[End]
...
So I think replacing text widget in functionally equivalent way is a lot of work.

Which specific problems did you have with existing Motif XmTextField widget? It works fine for me, but maybe my usage is less sophisticated.

**Comment by unixwork** [comment]
I've implemented middle mouse button, text select with double/triple click and some other shortcuts now. What is still missing is dragging selected text.

The main problem with XmTextField is, that is doesn't work sometimes. I still don't know why you couldn't type your national characters. I also know for sure that is doesn't work with a non-utf8 locale and I can't force people to use specific locales, because then some people would complain.

The other big limitation of XmTextField is, that with Xft you can't display all characters. Maybe people want to search for japanese, russian or what ever text that requires different fonts.

**Comment by dpeterc** [comment]
I can confirm that middle mouse button, text selection with double/triple click, [Home], [End], works.
There is still a small bug with middle mouse button paste; when new text is inserted, the "Find" button stays inactive (grayed out). The widget did not send the modify status, so find button remains disabled. Doing Ctrl+V works correctly.

As for the Motif XmTextField - it can display all characters with XFT, as long as the currently selected font support them.
On Mac, many fonts support wide range of characters. On Linux, a little less, but still showing mixed language text with XmTextField works OK with some fonts. See attached screenshot from my software, displaying Italian, Russian and Chinese in a Motif application using XmTextField with font "Noto Serif SC".
![fonts](https://user-images.githubusercontent.com/15419249/97816897-0be40900-1c99-11eb-9b79-bb891e6f7929.png)

**Comment by unixwork** [comment]
Ok, all problems with the new textfield should be resolved now.

I know that with the right font XmTextField can show a variety of characters. However it requires some configuration by the user, because I can't force them to use special fonts. People expect things to work right out of the box.

With the new widget, you have more freedom in choosing the font and it also works with all LANG settings. I plan to make it the default in the next release.

If everything is working now, I will close this ticket.

**Comment by dpeterc** [comment]
I confirm it works correctly now, thank you.
